plugin/etcd/plugin.go

```diff
@@ -0,0 +1,496 @@
+				Name:    "prefix",
+				Type:    "string",
+				Help:    "This is the input string for prefix based backup-restore.",
+				Title:   "Prefix",
+				Example: "starkandwayne",
+			},
+		},
+	}
+	fmt.Fprintf(os.Stderr, "etcd plugin starting up...\n")
+	plugin.Run(p)
+}
+
+type EtcdPlugin plugin.PluginInfo
+
+type EtcdConfig struct {
+	EtcdClient      *clientv3.Client
+	TimeoutDuration time.Duration
+	Overwrite       bool
+	Prefix          string
+}
+
+func (p EtcdPlugin) Meta() plugin.PluginInfo {
+	return plugin.PluginInfo(p)
+}
+
+func getEtcdConfig(endpoint plugin.ShieldEndpoint) (*EtcdConfig, error) {
+	etcdEndpoint, err := endpoint.StringValue("url")
+	if err != nil {
+		return nil, err
+	}
+
+	etcdUrls := strings.Split(etcdEndpoint, ",")
+
+	timeoutStr, err := endpoint.StringValueDefault("timeout", "2")
+	if err != nil {
+		return nil, err
+	}
+
+	timeout, err := strconv.Atoi(timeoutStr)
+	if err != nil {
+		return nil, err
+	}
+
+	duration := time.Duration(timeout) * time.Second
+
+	auth, err := endpoint.StringValueDefault("auth", "")
+	if err != nil {
+		return nil, err
+	}
+
+	username, err := endpoint.StringValueDefault("username", "")
+	if err != nil {
+		return nil, err
+	}
+
+	password, err := endpoint.StringValueDefault("password", "")
+	if err != nil {
+		return nil, err
+	}
+
+	clientCert, err := endpoint.StringValueDefault("client_cert", "")
+	if err != nil {
+		return nil, err
+	}
+
+	clientKey, err := endpoint.StringValueDefault("client_key", "")
+	if err != nil {
+		return nil, err
+	}
+
+	caCert, err := endpoint.StringValueDefault("ca_cert", "")
+	if err != nil {
+		return nil, err
+	}
+
+	overwrite, err := endpoint.BooleanValueDefault("overwrite", false)
+	if err != nil {
+		return nil, err
+	}
+
+	prefix, err := endpoint.StringValueDefault("prefix", "")
+	if err != nil {
+		return nil, err
+	}
+
+	tlsConfig := &tls.Config{}
+	if auth == "Certificate-Based Authentication" {
+		caCertPool := x509.NewCertPool()
+		check := caCertPool.AppendCertsFromPEM([]byte(caCert))
+		if check != true {
+			plugin.DEBUG("CA cert did't parse right.\n")
+		}
+
+		cert, err := tls.X509KeyPair([]byte(clientCert), []byte(clientKey))
+		if err != nil {
+			return nil, err
+		}
+
+		tlsConfig = &tls.Config{
+			Certificates: []tls.Certificate{cert},
+			RootCAs:      caCertPool,
+		}
+	}
+
+	cli, err := clientv3.New(clientv3.Config{
+		Endpoints:   etcdUrls,
+		DialTimeout: duration,
+		Username:    username,
+		Password:    password,
+		TLS:         tlsConfig,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &EtcdConfig{
+		EtcdClient:      cli,
+		TimeoutDuration: duration,
+		Overwrite:       overwrite,
+		Prefix:          prefix,
+	}, nil
+}
+
+//Validate user input
+func (p EtcdPlugin) Validate(endpoint plugin.ShieldEndpoint) error {
+	var (
+		s    string
+		b    bool
+		err  error
+		fail bool
+	)
+
+	roleAuth := "Role-Based Authentication"
+	certAuth := "Certificate-Based Authentication"
+	bothAuth := "Both"
```

Comment:
Make these global variables and use them everywhere.  Put them at the top of the file, and name them something like `RoleBasedAuthentication`, `CertificateAuthentication`, etc.